src/Skeleton.hs

```diff
@@ -9,132 +17,274 @@ module Skeleton
-  Gtk.set entry
+  entry <- liftGtk Gtk.entryNew
+  liftGtk $ Gtk.set entry
     [ Gtk.entryEditable := True
     , Gtk.widgetSensitive := True
     ]
 
-  -- text update event
-  void $ on entry Gtk.entryActivate $ do
-    -- This part simulates a worker thread doing a slow action then updating the state
-    -- in this implementation order is not guaranteed, which is a problem.
-    -- This is one of the key things we need in a reflex implementation.
-
-    -- clear current state since it is no longer valid
-    modifyIORef stateRef (\(GuiState k _) -> GuiState k Nothing)
+  -- This part simulates a worker thread doing a slow action then updating the state
+  -- in this implementation order is not guaranteed, which is a problem.
+  -- This is one of the key things we need in a reflex implementation.
 
+  -- Note: this pattern of using `newTriggerEvent` and then `on` to bind the
+  -- trigger will be extremely common, so we'll definitely want it to be
+  -- packaged up nicely for re-use
+  (startComputation, triggerStartComputation) <- newTriggerEvent
+  -- text update event
+  void $ liftGtk $ on entry Gtk.entryActivate $ do
+    putStrLn "entryActivate"
     -- get text
     txt <- Gtk.get entry Gtk.entryText :: IO String
+    triggerStartComputation txt
 
+  computationFinished <- performEventAsync $ ffor startComputation $ \txt done -> do
     -- run "slow" work in thread
-    void $ forkIO $ do
+    void $ liftIO $ forkIO $ do
       putStrLn $ "got text entry event (" <> txt <> "), simulating a worker thread..."
       threadDelay 2000000 -- 2 second delay
       putStrLn "worker thread finished doing work, updating state"
 
-      -- update state after work is done
-      Gtk.postGUISync $
-        modifyIORef stateRef (\(GuiState k _) -> GuiState k (Just txt))
+      done txt
+
+  let stateUpdate = leftmost
+        [ Nothing <$ startComputation
+        , Just <$> computationFinished
+        ]
 
-  pure entry
+  pure (entry, stateUpdate)
 
 
-makeCommitButton :: IORef GuiState -> (GuiState -> IO ()) -> IO Gtk.Button
-makeCommitButton stateRef outputAction = do
-  commitButton <- Gtk.buttonNewWithLabel "commit"
-  Gtk.widgetSetTooltipText commitButton
+makeCommitButton :: _ => m (Gtk.Button, Event t ())
+makeCommitButton = do
+  commitButton <- liftGtk $ Gtk.buttonNewWithLabel "commit"
+  liftGtk $ Gtk.widgetSetTooltipText commitButton
     (Just "SET ME SET ME GO HEAD DO IT COME ON SET ME")
 
-  _ <- on commitButton Gtk.buttonActivated $ do
+  (commit, triggerCommit) <- newTriggerEvent
+  _ <- liftGtk $ on commitButton Gtk.buttonActivated $ do
     putStrLn "commit pressed"
-    readIORef stateRef >>= outputAction
+    triggerCommit ()
+
+  pure (commitButton, commit)
+
+mainWindow :: _ => Event t GuiInputEvent -> (GuiState -> IO ()) -> m Gtk.Window
+mainWindow inputEvent outputAction = do
+  rec stateString <- holdDyn Nothing setStateString
+
+      -- Does this still need to be stateful at all?  I don't think we actually use
```

Comment:
It does not need to be stateful. I'd love to see it removed.

Reply:
https://github.com/ghorn/reflex-skeleton/pull/5